### ISSUE TITLE
examples/resnet50

### ISSUE THREAD

**Issue by fmscole** [google/jax#11537]
Has anyone trained resnet50 in example?
In a batch, I always get the same prediction value：
```
batch_size = 100
[ 56  56  56  56  56  56 153  56  56  56  56  56  56  56  56  56  56  56
  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56
  56  56  56  56  56  56 323  56  56  56  56  56  56  56  56  56  56  56
  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56
  56 695  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56  56
  56  56  56  56  56  56  56  56  56  56]
```

**Comment by fmscole** [comment]
Thanks !@mattjj
The learning rate should be set to 0.00001

**Comment by fmscole** [comment]
Should the loss function be modified ？
`loss=jnp.mean(-jnp.sum(logits * targets,axis=-1))`
such as ：
```
def loss_fn(params, batch):
    inputs, targets = batch
    logits = predict_fun(params, inputs)
    target_class =targets
    targets=jax.nn.one_hot(targets,1000)
    loss=jnp.mean(-jnp.sum(logits * targets,axis=-1))
    predicted_class = jnp.argmax(logits, axis=-1)
    acc_count=jnp.sum(predicted_class == target_class)
    return loss,(loss,acc_count,target_class,predicted_class)
```
The learning rate can be set to 0.001,not 0.1

**Comment by fmscole** [comment]
@mattjj 
According to your guidance，I retested the training process, like [Flax resnet50](https://github.com/google/flax/tree/main/examples/imagenet).

Here is the whole code.The dataset is ILSVRC2012_img_val.
```
import numpy.random as npr
import jax
import jax.numpy as jnp
from jax import jit, grad, random
from jax.example_libraries import stax
from jax.example_libraries.stax import (AvgPool, BatchNorm, Conv, Dense,
                                        FanInSum, FanOut, Flatten, GeneralConv,
                                        Identity, MaxPool, Relu, LogSoftmax,Softmax)

import ml_collections
def get_config():
  config = ml_collections.ConfigDict()
  config.model = 'ResNet50'
  config.dataset = 'ILSVRC2012_img_val'
  config.learning_rate = 0.1
  config.warmup_epochs = 5.0
  config.momentum = 0.9
  config.batch_size = 25
  config.num_epochs = 40.0
  config.log_every_steps = 100
  config.cache = False
  # config.half_precision = False
  config.num_train_steps =40000//25*40
  config.steps_per_eval =1000//25
  return config
config=get_config()

def ConvBlock(kernel_size, filters, strides=(2, 2)):
  ks = kernel_size
  filters1, filters2, filters3 = filters
  Main = stax.serial(
      Conv(filters1, (1, 1), strides), BatchNorm(), Relu,
      Conv(filters2, (ks, ks), padding='SAME'), BatchNorm(), Relu,
      Conv(filters3, (1, 1)), BatchNorm())
  Shortcut = stax.serial(Conv(filters3, (1, 1), strides), BatchNorm())
  return stax.serial(FanOut(2), stax.parallel(Main, Shortcut), FanInSum, Relu)

def IdentityBlock(kernel_size, filters):
  ks = kernel_size
  filters1, filters2 = filters
  def make_main(input_shape):
    return stax.serial(
        Conv(filters1, (1, 1)), BatchNorm(), Relu,
        Conv(filters2, (ks, ks), padding='SAME'), BatchNorm(), Relu,
        Conv(input_shape[3], (1, 1)), BatchNorm())
  Main = stax.shape_dependent(make_main)
  return stax.serial(FanOut(2), stax.parallel(Main, Identity), FanInSum, Relu)

def ResNet50(num_classes):
  return stax.serial(
      GeneralConv(('NHWC', 'HWIO', 'NHWC'), 64, (7, 7), (2, 2), 'SAME'),
      BatchNorm(), Relu, MaxPool((3, 3), strides=(2, 2)),
      ConvBlock(3, [64, 64, 256], strides=(1, 1)),
      IdentityBlock(3, [64, 64]),
      IdentityBlock(3, [64, 64]),
      ConvBlock(3, [128, 128, 512]),
      IdentityBlock(3, [128, 128]),
      IdentityBlock(3, [128, 128]),
      IdentityBlock(3, [128, 128]),
      ConvBlock(3, [256, 256, 1024]),
      IdentityBlock(3, [256, 256]),
      IdentityBlock(3, [256, 256]),
      IdentityBlock(3, [256, 256]),
      IdentityBlock(3, [256, 256]),
      IdentityBlock(3, [256, 256]),
      ConvBlock(3, [512, 512, 2048]),
      IdentityBlock(3, [512, 512]),
      IdentityBlock(3, [512, 512]),
      AvgPool((3, 3)), Flatten, Dense(num_classes), LogSoftmax)

rng_key = random.PRNGKey(0)  
num_classes = 1000
input_shape = (1,224, 224, 3)
init_fun, predict_fun = ResNet50(num_classes)  
_, init_params = init_fun(rng_key, input_shape)

base_learning_rate = config.learning_rate * config.batch_size / 256.
steps_per_epoch=40000//config.batch_size
def myschedule(steps):
    cosine_epochs = max(config.num_epochs - config.warmup_epochs, 1)
    warmup_decay_steps=config.warmup_epochs*steps_per_epoch
    cosine_decay_steps=cosine_epochs*steps_per_epoch
    return jnp.select([steps<config.warmup_epochs*steps_per_epoch,
                        steps<config.num_epochs* steps_per_epoch],
                    [base_learning_rate*steps/warmup_decay_steps,
                    base_learning_rate*(0.5*jnp.cos(jnp.pi*(steps-warmup_decay_steps)/cosine_decay_steps)+0.5+0.000001)],0.000001)

def loss_fn(params, batch):
    inputs, targets = batch
    logits = predict_fun(params, inputs)
    target_class =targets
    targets=jax.nn.one_hot(targets,1000)
    loss=jnp.mean(-jnp.sum(logits * targets,axis=-1))
    predicted_class = jnp.argmax(logits, axis=-1)
    acc_count=jnp.sum(predicted_class == target_class)
    return loss,(loss,acc_count,target_class,predicted_class)

@jit
def accuracy(params, batch):
    inputs, targets = batch
    target_class =targets
    predicted_class = jnp.argmax(predict_fun(params, inputs), axis=-1)
    return jnp.sum(predicted_class == target_class)
@jit
def update(i,params, updates, batch):
    g,ans=grad(loss_fn,has_aux=True)(params, batch)
    # moments=0.9    
    # updates=jax.tree_util.tree_map(lambda x,y:moments*x+(1-moments)*y,g,updates)    
    lr=myschedule(i)
    params=jax.tree_util.tree_map(lambda x,y: x-lr*y,params,g)
    return params,updates,ans,lr

if __name__ == "__main__":
  import numpy as np  
  from min_dataset import getDataLoader,numpy_collate
  import time
  train_ds =getDataLoader(batch_size=config.batch_size,collate_fn=numpy_collate)
  
  epochs=100  
  params=init_params
  updates=jax.tree_util.tree_map(lambda x:x*0.0,params)
  steps=0
  for e in range(epochs):
    test_ds =getDataLoader(batch_size=config.batch_size,is_train=False,collate_fn=numpy_collate)
    acc_count=0
    total=0
    for i,( x,y) in enumerate(test_ds):
      acc_count+=accuracy(params, ( x,y))
      total+=y.shape[0]
    print("")
    print("epoch:",e)
    print("test acc=",acc_count/total,i+1)
    
    aacc_countcc=0
    total=0
    begin=time.time()
    for i,( x,y) in enumerate(train_ds):
      steps+=1
      params,updates,(lv,batch_acc,target_class,predicted_class) ,lr= update(steps,params,updates, ( x,y))
      acc_count+=batch_acc
      total+=y.shape[0]

    print("train acc=",acc_count/total,time.time()-begin)
    print("lr=",lr)
    print("(",e,lv,",",batch_acc,")")    
    print(predicted_class)
```
result:
![jax_m_40](https://user-images.githubusercontent.com/8135902/180678046-fc4c34c5-26e7-4f91-ab5a-4724455c918f.jpg)
As part of the training, the data reading code is：
```
from torch.utils.data import Dataset,DataLoader,random_split
import torchvision.transforms as transforms
import numpy as np
from PIL import Image
class ImageNetBaseSet(Dataset):
    def __init__(self,data_root=r"/home/fms/data/ILSVRC2012_img_val"):               
        data_txt=data_root+r'/val_list.txt'
        self.transforms=transforms   

        files=open(data_txt,'r')
        content=files.readlines()
        files.close()
        self.image_paths=[]
        self.labels=[]
        for i in content:
            f,cls=i.split(" ")
            cls=int(cls)
            self.labels.append(cls)
            self.image_paths.append(data_root+"/images/"+f)                                   

    def __getitem__(self, index):
        image_file=self.image_paths[index]
        im=Image.open(image_file).convert("RGB")
        cls=self.labels[index] 
        return im,cls
    def __len__(self):
        return len(self.labels)

class ImageNetSet(Dataset):
    def __init__(self,baseset,transforms=None):               
        self.baseset=baseset
        self.transforms=transforms
    def __getitem__(self, index):
        im, cls=self.baseset.__getitem__(index)               
        if self.transforms:
            try:
                x=self.transforms(im)
            except:
                print(x.shape)
        else:
            im=im.resize((224,224))
            x=np.array(im)/255.0
        return x,cls
    def __len__(self):
        return self.baseset.__len__()

def numpy_collate(batch):
  if isinstance(batch[0], np.ndarray):
    return np.stack(batch)
  elif isinstance(batch[0], (tuple,list)):
    transposed = zip(*batch)
    return [numpy_collate(samples) for samples in transposed]
  else:
    if isinstance(batch, (tuple,list)):
      if not isinstance(batch[0], int):
        batch=[np.array(im).transpose(1,2,0) for im in batch]
    return np.array(batch)

basedataset=ImageNetBaseSet()
dataset40000, dataset10000 =random_split(basedataset, [40000, 10000])
normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                     std=[0.229, 0.224, 0.225])
trainset=ImageNetSet(dataset40000,transforms=transforms.Compose([
            transforms.RandomResizedCrop(224),
            transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
            normalize,
        ]))
valset=ImageNetSet(dataset10000,transforms=transforms.Compose([
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            normalize,
        ]))

def getDataLoader(batch_size=100,is_train=True,num_workers=16,shuffle=True,collate_fn=None):
    if is_train:
        training_generator = DataLoader(trainset, batch_size=batch_size, num_workers=num_workers,shuffle=shuffle,collate_fn=collate_fn)
        return training_generator
    else: 
        test_dataset1, test_dataset =random_split(valset, [9000, 1000])
        val_generator = DataLoader(test_dataset, batch_size=batch_size, num_workers=num_workers, shuffle=shuffle,collate_fn=collate_fn)
        return val_generator
```
For comparison, the result of flax resnet50 is：
![flax_40](https://user-images.githubusercontent.com/8135902/180678749-e887e3d1-e9a6-49cc-b5f3-005dc785720e.jpg)

**Comment by fmscole** [comment]
@mattjj 
Hello, after my test, the Jax version of resnet50 has the same performance as the flax version of resnet50. Please see the specific code [https://github.com/fmscole/jax_example](https://github.com/fmscole/jax_example). I think this code is simpler and more direct, more friendly to beginners, and more jaxic, such as using jax.tree_ util.tree_ Map. can you merge  this code into example?

**Comment by fmscole** [comment]
@mattjj 
Hello, after my test, the Jax version of resnet50 has the same performance as the flax version of resnet50. Please see the specific code [https://github.com/fmscole/jax_example](https://github.com/fmscole/jax_example). I think this code is simpler and more direct, more friendly to beginners, and more jaxic, such as using jax.tree_ util.tree_ Map. can you merge  this code into example?
